### DISCUSSION TITLE
Issue with setState

### DISCUSSION THREAD

**Discussion by SchattenOpa** [dowjones/react-dropdown-tree-select#485] [General]
First of all, 
I've seen other contributions and the "prevent-re-render"-solution given to previous issues.
However I don't think it works in my case so I'm opening a new issue here.

My problem is similar, when I call "setState" in the "onChange"-function my tree is collapsed and reset (for root-items they are still shown as "checked" though).

I don't create the data in my parent class for project-specific reasons, I retrieve it via a REST-Service and create my data-object in the "componentDidMount"-function in the child-component where the tree is located. 
(For simplicity-reasons I use the default-data in my example)
Creating the tree works great, as does navigating and selecting.
However as soon as I add the setState-function (which I need for saving the selection in the parent component for further use) I run into the described issue.
I can't do the comparison from the shouldComponentUpdate since I don't have anything to compare.
So I think I either would have to search my tree for the selected value and change its "checked"-value or update the whole tree in the state to the current (I assume this would also be necessary to keep it expanded?)

Is there any way to get the whole "data"-prop from the tree just like I get the currentNode? Or any other way to solve the issue?

I've created an example [HERE](https://codesandbox.io/s/issue-with-re-rendered-tree-in-react-dropdown-tree-select-f4zci?file=/DropdownContainer.js), it's simplified but the behavior is the same like in my own project.

**Comment by mrchief** [comment]
> Is there any way to get the whole "data"-prop from the tree

You already have the whole `data` because you're passing it to the tree in the first place. 

> I've created an example HERE, it's simplified but the behavior is the same like in my own project.

You're doing a `setState` and so it'll re-render the tree. AFAICT, using the prevent re-render HOC should fix your issue. Can you try bringing in the HOC from the examples and using that instead of the component directly?

**Reply by SchattenOpa** [reply]
> You already have the whole data because you're passing it to the tree in the first place.

I know, I only assumed the component would alter the props in some way, then I could maybe update my data with that to keep the changes after re-rendering.

> You're doing a setState and so it'll re-render the tree. AFAICT, using the prevent re-render HOC should fix your issue. Can you try bringing in the HOC from the examples and using that instead of the component directly?

I've tried that, but then the tree doesn't render at all because initially it is empty until I receive the data via rest.
Here's an example where I tried to include it:
[Example](https://codesandbox.io/s/issue-with-re-rendered-tree-in-react-dropdown-tree-select-modified-f4pt7?file=/index.js)

**Reply by mrchief** [reply]
I updated your [example](https://codesandbox.io/s/issue-with-re-rendered-tree-in-react-dropdown-tree-select-modified-forked-lkf3j). Moved around a few things and fixed your `isEqual` import which I think was the bigger culprit.

**Reply by SchattenOpa** [reply]
Thanks! 
I will see if this works in my original project

**Reply by SchattenOpa** [reply]
Thank you again, I have checked it now with my project and apparently it was indeed only my `isEqual` import which caused my problems.

**Reply by mrchief** [reply]
That's great news! Happy it got sorted out, thanks for using this library!
